src/sage/rings/species.py

```diff
@@ -0,0 +1,2341 @@
+        sage: a = lambda g, x: SetPartition([[g(e) for e in b] for b in x])
+        sage: _stabilizer_subgroups(S, X, a)
+        [Permutation Group with generators [(1,2), (1,3)(2,4)]]
+
+        sage: S = SymmetricGroup(8)
+        sage: X = SetPartitions(S.degree(), [3,3,2])
+        sage: _stabilizer_subgroups(S, X, a)
+        [Permutation Group with generators [(7,8), (6,7), (4,5), (1,3)(2,6)(4,7)(5,8), (1,3)]]
+
+        sage: S = SymmetricGroup(4)
+        sage: X = SetPartitions(S.degree(), 2)
+        sage: _stabilizer_subgroups(S, X, a)
+        [Permutation Group with generators [(1,4), (1,3,4)],
+         Permutation Group with generators [(1,3)(2,4), (1,4)]]
+
+    Let us keep 3 and 6 in separate blocks, and check that the
+    returned subgroups have the proper domains::
+
+        sage: S = SymmetricGroup([1,2,4,5,3,6]).young_subgroup([4, 2])
+        sage: X = [pi for pi in SetPartitions(6, [3,3]) if all(sum(1 for e in b if e % 3) == 2 for b in pi)]
+        sage: _stabilizer_subgroups(S, X, a)
+        [Permutation Group with generators [(1,2), (4,5), (1,4)(2,5)(3,6)]]
+
+    TESTS::
+
+        sage: _stabilizer_subgroups(SymmetricGroup(2), [1], lambda pi, H: H)
+        [Permutation Group with generators [(1,2)]]
+    """
+    to_gap = {x: i for i, x in enumerate(X, 1)}
+
+    X = set(X)  # because orbit_decomposition turns X into a set
+    g_orbits = [orbit_decomposition(X, lambda x: a(g, x))
+                for g in G.gens()]
+
+    gens = [PermutationGroupElement([tuple([to_gap[x] for x in o])
+                                     for o in g_orbit])
+            for g_orbit in g_orbits]
+    result = []
+    M = set(range(1, len(to_gap) + 1))
+    while M:
+        p = M.pop()
+        OS = libgap.OrbitStabilizer(G, p, G.gens(), gens)
+        result.append(PermutationGroup(gap_group=OS["stabilizer"],
+                                       domain=G.domain()))
+        M.difference_update(OS["orbit"].sage())
+    return result
+
+
+class MolecularSpecies(IndexedFreeAbelianMonoid):
+    """
+    The monoid of (multivariate) molecular species.
+    """
+    @staticmethod
+    def __classcall__(cls, *args, **kwds):
+        """
+        Normalize the arguments.
+
+        EXAMPLES::
+
+            sage: from sage.rings.species import MolecularSpecies
+            sage: MolecularSpecies("X,Y") is MolecularSpecies(["X", "Y"])
+            True
+
+            sage: MolecularSpecies("X,Y") == MolecularSpecies(["X", "Z"])
+            False
+        """
+        if isinstance(args[0], AtomicSpecies):
+            indices = args[0]
+        else:
+            assert "names" not in kwds or kwds["names"] is None
+            indices = AtomicSpecies(args[0])
+        category = Monoids().Commutative() & SetsWithGrading().Infinite()
+        return super().__classcall__(cls, indices,
+                                     prefix='', bracket=False,
+                                     category=category)
```

Comment:
This looks actually fishy to me.  I really would like to have `MolecularSpecies` only accept names as argument.  However, if I understand correctly, `MolecularSpecies.__init__` has to have the same signature, but apparently it must also have the same signature as `IndexedFreeAbelianMonoid.__classcall__` (or `__init__`?).

Reply:
No, you can explicitly skip that `__classcall__` and go to its base class's one through `UniqueRepresentation.__classcall__(cls, names)`. The `__init__` and `__classcall__` signatures (which can be different) only need to be able to take as input the final key/signature that is given to the `UniqueRepresentation`.